system/lib/fetch/emscripten_fetch.cpp

```diff
@@ -67,28 +67,46 @@ emscripten_fetch_t *emscripten_fetch(emscripten_fetch_attr_t *fetch_attr, const
 	}
 
 	emscripten_fetch_t *fetch = (emscripten_fetch_t *)malloc(sizeof(emscripten_fetch_t));
+	if (!fetch) return 0;
 	memset(fetch, 0, sizeof(emscripten_fetch_t));
 	fetch->id = globalFetchIdCounter++; // TODO: make this thread-safe!
 	fetch->userData = fetch_attr->userData;
-	fetch->url = strdup(url); // TODO: free
+	fetch->url = strdup(url);
 	fetch->__attributes = *fetch_attr;
-	fetch->__attributes.destinationPath = fetch->__attributes.destinationPath ? strdup(fetch->__attributes.destinationPath) : 0; // TODO: free
-	fetch->__attributes.userName = fetch->__attributes.userName ? strdup(fetch->__attributes.userName) : 0; // TODO: free
-	fetch->__attributes.password = fetch->__attributes.password ? strdup(fetch->__attributes.password) : 0; // TODO: free
+	fetch->__attributes.destinationPath = fetch->__attributes.destinationPath ? strdup(fetch->__attributes.destinationPath) : 0;
+	fetch->__attributes.userName = fetch->__attributes.userName ? strdup(fetch->__attributes.userName) : 0;
+	fetch->__attributes.password = fetch->__attributes.password ? strdup(fetch->__attributes.password) : 0;
 	if (fetch->__attributes.requestHeaders)
 	{
 		size_t headersCount;
 		for(headersCount = 0; fetch->__attributes.requestHeaders[headersCount]; ++headersCount);
 		const char** headers = (const char**)malloc((headersCount + 1) * sizeof(const char*));
+		if(!headers)
+		{
+			emscripten_fetch_free(fetch);
+			return 0;
+		}
 		for(size_t i = 0; i < headersCount; ++i)
+		{
 			headers[i] = strdup(fetch->__attributes.requestHeaders[i]);
+			if(!headers[i])
+			{
+				emscripten_fetch_free(fetch);
```

Comment:
Note that if one of the `strdup` calls (or even the allocation of `headers`, above) fails, `emscripten_fetch_free` is going to attempt to free `fetch->__attributes.requestHeaders`, which is what was passed in via `fetch_attr`, not the `requestHeaders` being allocating here.  This behavior is probably not the right thing.

Also, if the `strdup` calls fail, `headers` itself and any other strings that have been duplicated also need to be freed.

Reply:
If `strdup` fails, it should gracefully return 0 if I understand correctly. But when that happens, we indeed must make sure we have `strdup`ped all the fields, or zero them out manually so that we don't go and free user-submitted pointers. That's a very good catch.

Related to this, I was thinking that if one of the `strdup`s fail, it will mean that the fetch can fire off in "partial" state, with one or more of the strdupped fields not being present in the initiated fetch. That's something I'm thinking is something that can be left for a followup to address, since it will be a bit more complex, and I don't want to "string in" a collaborator to start to do extra work to fix everything in one go.